src/index.ts

```diff
@@ -1,130 +1,234 @@
+                  line => subscriber.next(line),
+                  () => {
+                    subscriber.complete();
+                    task.skip("Failed, falling back to AAPT...");
+                    fallBackToAapt = true;
+                  },
+                  () => subscriber.complete()
+                );
+              })
+          },
+          {
+            title: chalk`Encoding using AAPT {dim [fallback]}`,
+            skip: () => !fallBackToAapt,
+            task: () => apktool.encode(decodeDir, unsignedApkPath, false)
+          }
+        ])
     },
     {
-      title: 'Signing patched APK file',
-      task: () => new Observable(subscriber => {
-        (async () => {
-          await uberApkSigner.sign(unsignedApkPath)
-            .forEach(line => subscriber.next(line))
-
-          await fs.copyFile(
-            path.join(tmpDir, 'unsigned-aligned-debugSigned.apk'),
-            finishedApkPath,
-          )
-
-          subscriber.complete()
-        })()
-      }),
-    },
-  ]).run().catch(error => {
-    console.error(
-      chalk`\n  {red.inverse.bold  Failed! } An error occurred:\n\n`,
-      error.toString(),
-    )
-
-    process.exit(1)
-  })
+      title: "Signing patched APK file",
+      task: () =>
+        new Observable(subscriber => {
+          (async () => {
+            await uberApkSigner
+              .sign(unsignedApkPath)
+              .forEach(line => subscriber.next(line));
+
+            await fs.copyFile(
+              path.join(tmpDir, "unsigned-aligned-debugSigned.apk"),
+              finishedApkPath
+            );
+
+            subscriber.complete();
+          })();
+        })
+    }
+  ])
+    .run()
+    .catch(error => {
+      console.error(
+        chalk`\n  {red.inverse.bold  Failed! } An error occurred:\n\n`,
+        error.toString()
+      );
+
+      process.exit(1);
+    });
 
   if (usesAppBundle) {
     console.log(
       chalk`{yellow
   {inverse.bold  WARNING }
 
   This app seems to be using {bold Android App Bundle} which is dependent on Google Play.
-  As a result, installing it through an APK is probably {bold not going to work},
-  regardless of whether it has been patched by apk-mitm or not.}`
+  This feature is very {bold experimental} as a result {bold it might not work}`
+    );
+
+    console.log(chalk`
+    {dim ╭} {blue {bold apk-mitm} v${version}}
+    {dim ├ {bold apktool} ${apktool.version}
+    ╰ {bold uber-apk-signer} ${uberApkSigner.version}}
+    `);
+
+    console.log(chalk.dim(`  Using temporary directory:\n  ${tmpDir}\n`));
+
+    await new Listr(
+      [
+        {
+          title: "Unzipping App Bundle",
+          task: () => compression.unzip(apkPath, decodeDir)
+        },
+        {
+          title: "Generating a new signing key",
+          task: () => keytool.createCertificate()
+        },
+        {
+          title: "Decoding base.apk file",
+          task: (_, task) =>
+            new Observable(subscriber => {
```

Comment:
Instead of returning an `Observable` here, we should return the `Listr` instance responsible for modifying the APK file. This will result in Listr showing the "subtasks" (similar to what I'm currently doing [here](https://github.com/shroudedcode/apk-mitm/blob/03bc18cfc41a99acff2e7ddcdb5773d2193c2d50/src/index.ts#L69)) and more useful output in case something goes wrong.

Reply:
I've made some changes, what do you think ?

Reply:
Sorry, I think my comment was a little confusing. I meant that the `prepareApk` function should no longer be `async` and directly `return` (instead of `await`) [its `Listr` instance](https://github.com/shroudedcode/apk-mitm/blob/a15fbed5ea460a3853dfc6a7d0cd2bf558f56069/src/index.ts#L48) and not do *anything* after that (no `.run()`, no `.catch`, and no `console.log()`). Having it just return the `Listr` instance without even running it allows us to use it in two different ways:

- call `prepareApk` and `.run()` like we did before in case a single APK file is supplied
- *directly* return it [here](https://github.com/shroudedcode/apk-mitm/blob/a15fbed5ea460a3853dfc6a7d0cd2bf558f56069/src/index.ts#L166) (literally `task: () => prepareApk(...)`), so Listr can show the full "tree of tasks" while running them

To simplify this I would move any code that's not related to the core process of preparing the APK (including error and success logging — basically [these lines](https://github.com/shroudedcode/apk-mitm/blob/a15fbed5ea460a3853dfc6a7d0cd2bf558f56069/src/index.ts#L118-L130)) into the `cli.js` file.

Reply:
Got it ;) ✅